sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -153,6 +153,15 @@ object SetOperationPushDown extends Rule[LogicalPlan] with PredicateHelper {
         )
       )
 
+    // Adding extra Limit below UNION ALL if both left and right childs are not Limit.
+    // This heuristic is valid assuming there does not exist any Limit push-down rule.
+    case Limit(exp, Union(left, right))
+      if left.maxRows.isEmpty || right.maxRows.isEmpty =>
```

Comment:
Actually I think this branch is safe without this check, did I miss something here?